### DISCUSSION TITLE
A clean upgrade path from 16.1 to 17

### DISCUSSION THREAD

**Discussion by idc77** [keycloak/keycloak#10292] [Archive]
Reading the [v17 release announcement](https://www.keycloak.org/2022/02/keycloak-1700-released.html) you're sent all over the place and left alone, confused.

Can you please provide a clear upgrade path from 16.1 to 17?

**Comment by idc77** [comment]
In detail, I'm trying to upgrade from 16.1 wildfly to 17 quarkus, standalone.

- Database is Postgresql.
- http port is 4001

How do I upgrade?

**Comment by idc77** [comment]
Postgres database is found at conf/keycloak.conf
http port maybe found in conf/quarkus.properties as `quarkus.http.port=4001` untested.

What do I set the hostname to when running behind nginx? localhost or the actual hostname?

**Comment by idc77** [comment]
What is kcadm.sh? Why is everything just so vague and undocumented?

**Comment by idc77** [comment]
```
ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) Error details:: java.lang.IllegalStateException: Key material not provided to setup HTTPS. Please configure your keys/certificates or start the server in development mode.
```

I don't want to listen on https, only http. How can I disable this crap?

**Comment by idc77** [comment]
What does http.enabled=true in conf/keycloak.conf do?
WHERE IS THE DOCUMENTATION!!!

**Comment by idc77** [comment]
I think it's best to dump keycloak for a better solution.
You can't just fucking change everything without documenting it and set the user all over the place to little vague bits of information.
What the FUCK

**Comment by DGuhr** [comment]
Hey @idc77. Sad to hear you have a bad experience here, we tried our best to not let people down. 

You can find all docs available for Keycloak 17 server installation in the [server guides](https://www.keycloak.org/guides#server) section, covering e.g. [how to configure your database and database vendor](https://www.keycloak.org/server/db), as mentioned in the announcement.

Also, the [general configuration guide](https://www.keycloak.org/server/configuration) explaining some core concepts, e.g. the different config sources from keycloak.conf over env variables and the CLI, and the [all config reference](https://www.keycloak.org/server/all-config) may be of interest.

Alternatively, you could look at the extensive help we created in the CLI by invoking e.g. `kc.sh start-dev --help` - there you'd see that an option `--http-port` is available to change the port to your desired one.

That said, to try out keycloak on postgres with port 4001, it's simply put:

`kc.sh|bat start-dev --db=postgres --http-port=4001 --db-username=<name> --db-password=<pw>`. The default database implementation assumes you run postgres on localhost:5432, if not, you could use `--db-url` to change the whole jdbc url, or to only change parts, `--db-url-hostname` etc. (please consult the help /( all config page)

`http-enabled` actually enables or disables http-traffic, as the name might suggest. You can find more in the [all config reference](https://www.keycloak.org/server/all-config)
Hope this helps.

**Reply by audioqube** [reply]
Hello, I'm trying to migrate from keycloak 16.1.1 where I developed a UserStorageProvider that access an external Oracle DB for legacy users. Before I used an XA datasource configured inside wildfly and now I don't know how to create an additional Datasource for Oracle database in quarkus. I tried to configure a datasource using the quarkus.properties file but when i start keycloak 17 in my local PC with normal Openjdk I have many errors. It seems that the default Postgres Database for keycloak cannot access the db anymore because it cannot read the password !! Please help me.

Thank you in advance

**Reply by DGuhr** [reply]
Hey @audioqube - there were multiple issues around for the external datasource / userstorageprovider. Good news: With the next keycloak version / latest main, you'll get it working again :) 

Here you see the merged PR in main:
https://github.com/keycloak/keycloak/pull/10581

And here the not-yet merged PR in quickstarts for the new quickstart:
https://github.com/keycloak/keycloak-quickstarts/pull/302

So if you want to try it out, you can build latest main and do it like in the quickstart. Feel free to provide feedback in form of creating a Q/A discussion or an issue when something is not working for you, but better not in a subsequent comment in other threads, these get easily overlooked at least for me ;)

**Comment by drewfreyling** [comment]
The links to the Quarkus miggration guide are busted everywhere - migration/migrating-to-quarkus[Quarkus Migration Guide].

**Reply by audioqube** [reply]
DO you know how to create a separate datasource in order to access external user data from inside custom user storage provider?

**Comment by mazbeh** [comment]
@DGuhr hello and thanks for your tipps !
i am trying to update keycloak from version 12 to version 17. What should i do, so that i can have alle of my realms and clients from the last version into the new version?
i have already followed some suggestions in internet but i could not do it ...

could you please give me some tipps?

Thanks in advance !
